### DISCUSSION TITLE
Feature request : regenerate at any level of a conversation

### DISCUSSION THREAD

**Discussion by mlier** [open-webui/open-webui#3045] [Ideas]
**Is your feature request related to a problem? Please describe.**
There is a regenerate button after each conversation. 

<img width="1020" alt="regenerateoneconversation" src="https://github.com/open-webui/open-webui/assets/4807436/a55752bd-f95d-40cf-bcd8-8f11b0290785">

But that button disappear if you continue the conversation.
So one cannot continu the conversation from the upper part of the tree conversation.
![noregeneratetwoconversation](https://github.com/open-webui/open-webui/assets/4807436/ca018607-451b-4ea1-864b-0e883f0b3e6a)

**Describe the solution you'd like**
Add regenerate button at any level of the conversation.

**Comment by justinh-rahb** [comment]
**Editing** a user message mid-thread typically creates a new branch of conversation without the previous subsequent messages in the context, as subsequent turns are no longer related to the original conversation node. Likewise, _regenerating_ a message mid-thread is not feasible due to the nature of how chat completion requests work. This functionality would be more appropriate within the Playground workflow, which reinforces the need for a button to send chats to the Playground:
- #2565

**Comment by mlier** [comment]
It is already possible to do it today even if it is horrible from UX point of view. 
Each time you make a request and have a response, do a regenerate.
For instance, you will have the following conversation, each node is regenerated immediatly.

![1oneconversation_regenerateeachnode](https://github.com/open-webui/open-webui/assets/4807436/1a7b8cb9-4daf-4b8a-b7f6-261c91ebce7b)

Now a new conversation is started from the second request "Re Hello".
It is possible to add a new branch of conversation from that place : "Hello Paris" and "Hello Madrid"

![2newconversation_fromsecondnode](https://github.com/open-webui/open-webui/assets/4807436/f35d6f42-0185-4467-a7f1-8773875ad0b5)

That solution is not optimal : node duplication, not used.
The good solution : regenerate node in the middle of a conversation when the user ask it.

 It would be useful because the user can unvalidate (Bad Response) parts of the conversation and continu the conversation from an upper node already validated (Good Response).

**Comment by barryd15** [comment]
@mlier 
I have 2 workarounds, both awkward, but less so than the above example where you prepare nodes in advance:
1. Click "Edit" on a response node, click "Save as Copy", then click "Regenerate" on the newly-created leaf duplicate response node.
2. Click "Edit" on the response node's parent request node, click "Send" without modifying the message (though this creates an unnecessary duplicate request node).

FYI, I don't know if these workarounds were possible when this issue was first reported.

I also would like to see the "Regenerate" icon on every response node, not just the leaf nodes.
I was going to raise a discussion about it, asking why the design is like this, before I found this one.

@justinh-rahb 
If I understand your objection right (and I'm not sure I do), it must be a performance issue you are raising, because cached information cannot be used.
Otherwise, how is it possible that we can use the above workarounds to branch in the middle of a chat?
And how is it possible that we can take an old chat, restart Ollama, and resume the chat at the endpoint?

**Reply by justinh-rahb** [reply]
The relevant code and functionality has changed significantly since I wrote that.

**Reply by barryd15** [reply]
That's good to hear.  It seems like the requested change will be much more viable in the current version of the code then.

**Reply by barryd15** [reply]
Just want to note here that this appears to be in place in v0.4.4 already, though I didn't yet figure out exactly at what point it was added.
